DOCSP-35943: write operations reorg (#3275)
jenssegers/laravel-mongodb: docs/fundamentals/write-operations/modify.txt

   :end-before: end model update multiple

When the ``update()`` method succeeds, the operation returns the number of
documents updated.

If the retrieve part of the call does not match any documents in the
collection, the {+odm-short+} returns the following error:

.. code-block:: none
   :copyable: false

   Error: Call to a member function update() on null

.. _laravel-modify-documents-upsert:

Update or Insert in a Single Operation
--------------------------------------

An **upsert** operation lets you perform an update or insert in a single
operation. This operation streamlines the task of updating a document or
inserting one if it does not exist.

Starting in v4.7, you can perform an upsert operation by using either of
the following methods:

- ``upsert()``: When you use this method, you can perform a **batch
  upsert** to change or insert multiple documents in one operation.

- ``update()``: When you use this method, you must specify the
  ``upsert`` option to update all documents that match the query filter
  or insert one document if no documents are matched. Only this upsert method
  is supported in versions v4.6 and earlier.

Upsert Method
~~~~~~~~~~~~~

> The ``upsert(array $values, array|string $uniqueBy, array|null

Reviewer: S: since you list out all the parameters, I think you can remove them here
```suggestion
The ``upsert()
```

$update)`` method accepts the following parameters:

- ``$values``: Array of fields and values that specify documents to update or insert.
- ``$uniqueBy``: List of fields that uniquely identify documents in your
  first array parameter.
- ``$update``: Optional list of fields to update if a matching document
  exists. If you omit this parameter, the {+odm-short+} updates all fields.

> To specify an upsert in the ``upsert()`` method, set parameters

Reviewer: S:
```suggestion
To specify an upsert in the ``upsert()`` method, pass the required parameters
```

      }

.. _laravel-modify-documents-arrays:

Update Arrays in a Document
---------------------------

In this section, you can see examples of the following operations that
update array values in a MongoDB document:

- :ref:`Add values to an array <laravel-modify-documents-add-array-values>`
- :ref:`Remove values from an array <laravel-modify-documents-remove-array-values>`
- :ref:`Update the value of an array element <laravel-modify-documents-update-array-values>`

These examples modify the sample document created by the following insert
operation:

.. literalinclude:: /includes/fundamentals/write-operations/WriteOperationsTest.php
   :language: php
   :dedent:
   :start-after: begin array example document
   :end-before: end array example document

.. _laravel-modify-documents-add-array-values:

Add Values to an Array Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This section shows how to use the ``push()`` method to add values to an array
in a MongoDB document. You can pass one or more values to add and set the
optional parameter ``unique`` to ``true`` to skip adding any duplicate values
in the array. The following code example shows the structure of a ``push()``
method call:

> .. code-block:: none

Reviewer: S:
```suggestion
.. code-block:: php
```

          <field name>,
          [<values>], // array or single value to add
          unique: true); // whether to skip existing values

The following example shows how to add the value ``"baroque"`` to
the ``genres`` array field of a matching document. Click the
:guilabel:`{+code-output-label+}` button to see the updated document:

.. io-code-block::

   .. input:: /includes/fundamentals/write-operations/WriteOperationsTest.php
      :language: php
      :dedent:
      :start-after: begin model array push
      :end-before: end model array push

   .. output::
      :language: json
      :visible: false

      {
        "_id": "660eb...",
        "performer": "Mitsuko Uchida",
        "genres": [
            "classical",
            "dance-pop",

        ],
        "updated_at": ...,
        "created_at": ...
      }

.. _laravel-modify-documents-remove-array-values:

Remove Values From an Array Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This section shows how to use the ``pull()`` method to remove values from
an array in a MongoDB document. You can pass one or more values to remove
from the array. The following code example shows the structure of a
``pull()`` method call:

> .. code-block:: none

Reviewer: S:
```suggestion
.. code-block:: php
```

Reviewer: applies to the other syntax code examples too
